software/controller/lib/core/commfail_alarm.cpp

```diff
@@ -0,0 +1,104 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <stdint.h>
+#include "debug.h"
+#include "hal.h"
+#include "commfail_alarm.h"
+#include "comms.h"
+
+// This count checks how many times alarm is triggered before it raises alarm
+// for medium & low priority alarms, where as alarm will be raised 
+// immediately for high priority alarms
+static constexpr uint8_t kTriggerCount = 100;
+
+CommFailAlarm commfailalarm;
```

Comment:
Now that there isn't an actual callback as such, it is I think even more clear that there is no need for a global variable. This should be somehow initialized in `main`

Reply:
Declared "CommFailAlarm alarm" class variable in main, "extern CommFailAlarm alarm" at commfail_alarm.cpp. 
If not declared as extern & pass the class variable alarm to commfail_alarm functions it is not working, as a local class variable is being created for every function call (alarm.Handler from main.cpp).
Please suggest on this.

Reply:
But this is actually the problem. The functions should be manipulating the member variables of the object they are bound to, and not those of a global instance. There should be no global instance. The CommFailAlarm object should be initialized **INSIDE** a function of `main.cpp` and not as a global variable. Then you will have no need for this magic.

Reply:
May be I need to understand more object bound concept